src/components/Filters/index.tsx

```diff
@@ -66,482 +30,293 @@ const FilterWrapper = styled(Box)`
+	const filters = restProps.filters ? restProps.filters : [];
+	const views = restProps.views ? restProps.views : [];
+
+	const flatSchema = SchemaSieve.flattenSchema(schema);
+	const flatViews = flattenViews(views);
+	const flatFilters = filters.map(filter => SchemaSieve.flattenSchema(filter));
+	const shouldHideLabel = useBreakpoint(compact || [false]);
+
+	const [filtersState, setFiltersState] = React.useState<FiltersState>({
+		showModal: false,
+		searchString: '',
+		editingFilter: null,
+		filters: flatFilters,
+		views: flatViews,
+		schema: flatSchema,
+		edit: [],
+	});
+
+	React.useEffect(() => {
+		filtersState.edit.push(SchemaSieve.getCleanEditModel(filtersState.schema));
+	}, []);
+
+	React.useEffect(() => {
+		setFiltersState(prevState => ({
+			...prevState,
+			schema: SchemaSieve.flattenSchema(schema),
+		}));
+	}, [schema]);
 
-	public setEditOperator(operator: string, index: number) {
-		const currentEdit = this.state.edit.slice();
-		const item = currentEdit[index];
-		currentEdit.splice(index, 1, { ...item, operator });
-		this.setState({
-			edit: currentEdit,
-		});
-	}
+	React.useEffect(() => {
+		setFiltersState(prevState => ({
+			...prevState,
+			filters: filters.map(filter => SchemaSieve.flattenSchema(filter)),
+		}));
+	}, [useDeepCompare(filters)]);
 
-	public setEditValue(value: string, index: number) {
-		const currentEdit = this.state.edit.slice();
-		const item = currentEdit[index];
-		currentEdit.splice(index, 1, { ...item, value });
-		this.setState({
-			edit: currentEdit,
-		});
-	}
+	React.useEffect(() => {
+		setFiltersState(prevState => ({
+			...prevState,
+			views: flattenViews(views),
+		}));
+	}, [useDeepCompare(views)]);
 
-	public emitFilterUpdate() {
-		if (!this.props.onFiltersUpdate) {
+	React.useEffect(() => {
+		if (!onFiltersUpdate) {
 			return;
 		}
 
-		this.props.onFiltersUpdate(
-			this.state.filters.map(filter => SchemaSieve.unflattenSchema(filter)),
+		onFiltersUpdate(
+			filtersState.filters.map(filter => SchemaSieve.unflattenSchema(filter)),
 		);
-	}
+	}, [useDeepCompare(filtersState.filters)]);
+
+	React.useEffect(() => onViewsUpdate && onViewsUpdate(filtersState.views), [
+		useDeepCompare(filtersState.views),
+	]);
 
-	public addFilter() {
-		const $id = this.state.editingFilter;
-		const { schema } = this.state;
-		const filter = SchemaSieve.createFilter(schema, this.state.edit);
-		const currentFilters = this.state.filters;
+	const addFilter = (edit: EditModel[]) => {
+		const $id = filtersState.editingFilter;
+		const { schema } = filtersState;
+		const filter = SchemaSieve.createFilter(schema, edit);
 
-		let filters: JSONSchema6[];
+		let currentFilters: JSONSchema6[];
 
 		if (!!$id) {
-			const matchIndex = findIndex(currentFilters, { $id });
-			currentFilters.splice(matchIndex, 1, filter);
-			filters = currentFilters.slice();
+			const matchIndex = findIndex(filtersState.filters, { $id });
+			filtersState.filters.splice(matchIndex, 1, filter);
```

Comment:
This mutates the filters, which means the component won't rerender. Let's use immutable functions instead.